### ISSUE TITLE
Expose kernels to 3rd party extensions

### ISSUE THREAD

**Issue by DonJayamanne** [microsoft/vscode-jupyter#5532]
This is a hackathon project, and the some of the original requirements can be found here https://github.com/microsoft/vscode-jupyter/issues/3998

**Goal: Expose kernel to 3rd party extension authors so they can build on top of Jupyter extension**

**Possible extensions (prototyped):**
* [x] Other extensions can now execute code against kernels
	* User will be asked to allow such exensions to execute on behalf of them
	* 3rd party extensions can install missing packages, & the like (prototyped)
	* 3rd party extensions can inject kernel exensions (prototyped)
	* 3rd party extensions can intercept kernel execution requests
* [ ] SOS Jupyter Kernel (partially working)
	* We need cell pre-execute
* [x] Support functionality found in Jupyter Sidecar widget
	* Allow user to open widgets outside the notebook.
* [x] Custom variable explorer
	* Support for other languages (currently working, Python, R, Scala, C#, F#)
	* Support for variable visualizers (currently working for Python)
	* Other extensions can build on top and support more visualizers for each language
	* TOOD: 
		* Provide extensibility for other extensions to provide details of each variable so that the tree view can be augmented (e.g. drilling down into details of each variable)
		*  Ability to drill down into dict, dataframes in the variable viewer (partially protototyped)
* [x] Better plot viewer
	* Custom visualizer for variable, to open a plot in a seprate view (use ipympl)
	* User doesn't need to isntall ipympl (source shipped with 3rd party extension)
* [x] Custom data frameviewer
	* Basically similar to rest, display dataframe variable with the qgrid widget
	* Ship qgrid widget with 3rd party extension
	* Partially working
* [ ] Open altair (vega) plots with vega viewers
	* Convert vega into JSON, use 3rd party extensions to view the vega json data
	* I.e. more custom visualizers, just built on top of this new functionality

![Screen Shot 2021-04-15 at 11 29 23](https://user-images.githubusercontent.com/1948812/114951115-4ab1e980-9e09-11eb-9131-0db2f1a799a6.png)
![Screen Shot 2021-04-15 at 13 24 11](https://user-images.githubusercontent.com/1948812/114951110-47b6f900-9e09-11eb-9835-c21e3d233639.png)
![Screen Shot 2021-04-15 at 12 51 03](https://user-images.githubusercontent.com/1948812/114951111-484f8f80-9e09-11eb-8584-5aa912a08494.png)

**Comment by tanhakabir** [comment]
I want to support terminal commands for TS notebooks and JS notebooks running in vscode. Currently these kernels don't have support for terminal commands. Vscode however has apis for running terminal commands.

It would nice to have my extension intercept the cell before it goes to the respective kernel for any terminal commands.

**Comment by DonJayamanne** [comment]
Todo:
* [ ] Publish *.d.ts files
* [ ] Publish sample extension(s)
* [ ] Documentation (Wiki):
	* How to enroll into API
	* How to debug/test
	* Guidelines
* [ ] Release management
	* How do we notify changes (since we have a list of who can use this API, we can control this)
	* E.g. what if user installs an extension that uses an old version of the API (do we fail silently, do we display a message, if so who - extension or us etc.)

**Comment by DonJayamanne** [comment]
Closing this for now, a sample usage of the API can be found here https://github.com/microsoft/vscode-jupyter-powertoys/
